src/cargo/util/diagnostic_server.rs

```diff
@@ -258,14 +258,23 @@ impl RustfixDiagnosticServer {
 
     fn run(self, on_message: &dyn Fn(Message), done: &AtomicBool) {
         while let Ok((client, _)) = self.listener.accept() {
-            let client = BufReader::new(client);
-            match serde_json::from_reader(client) {
-                Ok(message) => on_message(message),
-                Err(e) => warn!("invalid diagnostics message: {}", e),
-            }
             if done.load(Ordering::SeqCst) {
                 break;
             }
+            let mut client = BufReader::new(client);
+            let mut s = String::new();
+            if let Err(e) = client.read_to_string(&mut s) {
+                warn!("diagnostic server failed to read: {}", e);
+            } else {
+                match serde_json::from_str(&s) {
```

Comment:
Why the change from `serde_json::from_reader`?  For the more precise warning message?

Reply:
`from_reader` consumes the reader, and the drops it which closes the connection to the client too early.

Reply:
Ah right.